File: src/year2024/day6/day6.go
// [Day 6: Guard Gallivant](https://adventofcode.com/2024/day/6)

package main

import (
	"fmt"
	"log"
	"os"
	"strings"
	"time"
)

// Coord represents a 2D coordinate.
type Coord struct {
	X, Y int
}

var (
	WEST  = Coord{X: -1, Y: 0}
	EAST  = Coord{X: 1, Y: 0}
	NORTH = Coord{X: 0, Y: -1}
	SOUTH = Coord{X: 0, Y: 1}
)

type Grid struct {
	data   []byte
	width  int
	height int
}

func parseGrid(input string) Grid {
	lines := strings.Split(strings.TrimSpace(input), "\n")
	if len(lines) == 0 {
		return Grid{}
	}

	width := 0
	for _, line := range lines {
		if len(line) > width {
			width = len(line)
		}
	}
	height := len(lines)

	var gridData []byte
	for _, line := range lines {
		gridData = append(gridData, []byte(line)...)
		for len(line) < width {
			gridData = append(gridData, ' ')
			line += " "
		}
	}

	return Grid{
		data:   gridData,
		width:  width,
		height: height,
	}
}

func (g Grid) get(x, y int) byte {
	if x < 0 || x >= g.width || y < 0 || y >= g.height {
		return 0
	}
	return g.data[y*g.width+x]
}

type state struct {
	pos Coord
	dir Coord
}

type Puzzle struct {
	grid  Grid
	start Coord
}

func newPuzzle(data string) *Puzzle {
	grid := parseGrid(data)

	var start Coord
	found := false
	for y := 0; y < grid.height; y++ {
		for x := 0; x < grid.width; x++ {
			if grid.get(x, y) == '^' {
				start = Coord{X: x, Y: y}
				found = true
				break
			}
		}
		if found {
			break
		}
	}

	return &Puzzle{grid: grid, start: start}
}

func (p *Puzzle) moveGuard(x, y *int, direction *Coord, obstruction Coord) bool {
	// obstruction with MAX coordinates to signify there is no obstruction
	noObstruction := Coord{X: 1<<31 - 1, Y: 1<<31 - 1}

	switch *direction {
	case EAST:
		if *x == 0 {
			return true
		}
		nextX, nextY := *x-1, *y
		if p.grid.get(nextX, nextY) == '#' || (nextX == obstruction.X && nextY == obstruction.Y && obstruction != noObstruction) {
			*direction = NORTH
		} else {
			*x = nextX
		}

	case WEST:
		if *x == p.grid.width-1 {
			return true
		}
		nextX, nextY := *x+1, *y
		if p.grid.get(nextX, nextY) == '#' || (nextX == obstruction.X && nextY == obstruction.Y && obstruction != noObstruction) {
			*direction = SOUTH
		} else {
			*x = nextX
		}

	case NORTH:
		if *y == 0 {
			return true
		}
		nextX, nextY := *x, *y-1
		if p.grid.get(nextX, nextY) == '#' || (nextX == obstruction.X && nextY == obstruction.Y && obstruction != noObstruction) {
			*direction = WEST
		} else {
			*y = nextY
		}

	case SOUTH:
		if *y == p.grid.height-1 {
			return true
		}
		nextX, nextY := *x, *y+1
		if p.grid.get(nextX, nextY) == '#' || (nextX == obstruction.X && nextY == obstruction.Y && obstruction != noObstruction) {
			*direction = EAST
		} else {
			*y = nextY
		}
	}

	return false
}

func (p *Puzzle) part1() int {
	x, y := p.start.X, p.start.Y
	direction := NORTH
	leave := false

	seen := make(map[Coord]bool)
	obstruction := Coord{X: 1<<31 - 1, Y: 1<<31 - 1}

	for !leave {
		seen[Coord{X: x, Y: y}] = true
		leave = p.moveGuard(&x, &y, &direction, obstruction)
	}

	return len(seen)
}

func (p *Puzzle) part2() int {
	xy := p.start
	direction := NORTH
	leave := false
	obstruction := Coord{X: 1<<31 - 1, Y: 1<<31 - 1}

	normalWalk := make(map[Coord]bool)

	for !leave {
		normalWalk[xy] = true
		leave = p.moveGuard(&xy.X, &xy.Y, &direction, obstruction)
	}

	stuck := 0

	for y := 0; y < p.grid.height; y++ {
		for x := 0; x < p.grid.width; x++ {
			xy := Coord{X: x, Y: y}
			c := p.grid.get(x, y)

			// optimization: if the guard never walks to this position,
			// an obstruction cannot deviate him
			if !normalWalk[xy] {
				continue
			}

			if c == '.' {
				// can choose this position for the obstruction
				obstruction := Coord{X: x, Y: y}

				xy := p.start
				direction := NORTH
				leave := false
				seen := make(map[state]bool)

				for !leave {
					state := state{pos: xy, dir: direction}
					if seen[state] {
						// same pos, same direction : the guard is stuck
						stuck++
						break
					}
					seen[state] = true

					leave = p.moveGuard(&xy.X, &xy.Y, &direction, obstruction)
				}
			}
		}
	}

	return stuck
}

func solve(data []byte) (int, int) {
	input := string(data)
	puzzle := newPuzzle(input)
	return puzzle.part1(), puzzle.part2()
}

func main() {
	filename := "input.txt"
	elapsed := false

	for i := 1; i < len(os.Args); i++ {
		arg := os.Args[i]
		if arg == "--elapsed" {
			elapsed = true
		} else if !strings.HasPrefix(arg, "-") {
			filename = arg
		}
	}

	data, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}

	if elapsed {
		start := time.Now()
		result1, result2 := solve(data)
		duration := time.Since(start)

		fmt.Println(result1)
		fmt.Println(result2)
		fmt.Printf("elapsed: %f ms\n", duration.Seconds()*1000.0)
	} else {
		result1, result2 := solve(data)

		fmt.Println(result1)
		fmt.Println(result2)
	}
}

File: src/year2024/day6/day6.rs
//! [Day 6: Guard Gallivant](https://adventofcode.com/2024/day/6)

use aoc::Coord;
use rayon::prelude::*;
use rustc_hash::FxHashSet;

type Grid = aoc::Grid<u8>;

struct Puzzle {
    grid: Grid,
    start: Coord,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        let grid = Grid::parse(data);
        let start = grid.iter().find(|(_, c)| c == &&b'^').unwrap().0;

        Self { grid, start }
    }

    /// Solve part one.
    fn part1(&self) -> usize {
        let width = self.grid.width();
        let height = self.grid.height();
        let Coord { mut x, mut y } = self.start;
        let mut dir = 0;

        let mut seen = FxHashSet::default();

        loop {
            seen.insert((x, y));

            let (next_x, next_y) = match dir {
                0 => (x, y - 1), // North
                1 => (x + 1, y), // East
                2 => (x, y + 1), // South
                3 => (x - 1, y), // West
                _ => unreachable!(),
            };

            if next_x < 0 || next_x >= width || next_y < 0 || next_y >= height {
                break;
            }

            if self.grid[(next_x, next_y)] == b'#' {
                dir = (dir + 1) % 4;
            } else {
                x = next_x;
                y = next_y;
            }
        }

        seen.len()
    }

    /// Solve part two.
    fn part2(&self) -> u32 {
        let width = self.grid.width();
        let height = self.grid.height();
        let Coord {
            x: start_x,
            y: start_y,
        } = self.start;

        // Trace original path to find candidates
        let mut candidates = Vec::new();

        let mut x = start_x;
        let mut y = start_y;
        let mut dir = 0;
        let mut seen = FxHashSet::default();

        #[allow(clippy::cast_sign_loss)]
        let visited_idx =
            |x: i32, y: i32, dir: i32| -> usize { ((y * width + x) * 4 + dir) as usize };
        let visited_size = visited_idx(0, height, 0);

        loop {
            if seen.insert((x, y)) {
                // Add as candidate if it's not the start position and valid
                if (x, y) != (start_x, start_y) {
                    candidates.push((x, y));
                }
            }

            let (next_x, next_y) = match dir {
                0 => (x, y - 1), // North
                1 => (x + 1, y), // East
                2 => (x, y + 1), // South
                3 => (x - 1, y), // West
                _ => unreachable!(),
            };

            if next_x < 0 || next_x >= width || next_y < 0 || next_y >= height {
                break;
            }

            if self.grid[(next_x, next_y)] == b'#' {
                dir = (dir + 1) % 4;
            } else {
                x = next_x;
                y = next_y;
            }
        }

        // Parallel optimization:
        // Use map_init to create a thread-local visited buffer and run_id.
        // This avoids reallocating the buffer for every candidate.
        candidates
            .par_iter()
            .map_init(
                || (vec![0u16; visited_size], 0u16),
                |(visited, run_id), &(ox, oy)| {
                    *run_id = run_id.wrapping_add(1);
                    if *run_id == 0 {
                        visited.fill(0);
                        *run_id = 1;
                    }
                    let current_run_id = *run_id;

                    let mut x = start_x;
                    let mut y = start_y;
                    let mut dir = 0; // North

                    loop {
                        let idx = visited_idx(x, y, dir);
                        if visited[idx] == current_run_id {
                            return 1; // Stuck
                        }
                        visited[idx] = current_run_id;

                        let (next_x, next_y) = match dir {
                            0 => (x, y - 1), // North
                            1 => (x + 1, y), // East
                            2 => (x, y + 1), // South
                            3 => (x - 1, y), // West
                            _ => unreachable!(),
                        };

                        if next_x < 0 || next_x >= width || next_y < 0 || next_y >= height {
                            return 0; // Outside of map
                        }

                        if self.grid[(next_x, next_y)] == b'#' || (next_x == ox && next_y == oy) {
                            dir = (dir + 1) % 4;
                        } else {
                            x = next_x;
                            y = next_y;
                        }
                    }
                },
            )
            .sum()
    }
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (usize, u32) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 41);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 6);
    }
}
